planner/core/logical_plans.go

```diff
@@ -211,14 +217,20 @@ func (p *LogicalJoin) columnSubstitute(schema *expression.Schema, exprs []expres
 // `OtherConditions` by the result of extract.
 func (p *LogicalJoin) AttachOnConds(onConds []expression.Expression) {
 	eq, left, right, other := p.extractOnCondition(onConds, false, false)
+	p.AppendJoinConds(eq, left, right, other)
+}
+
+// AppendJoinConds appends new join conditions.
+func (p *LogicalJoin) AppendJoinConds(eq []*expression.ScalarFunction, left, right, other []expression.Expression) {
 	p.EqualConditions = append(eq, p.EqualConditions...)
 	p.LeftConditions = append(left, p.LeftConditions...)
 	p.RightConditions = append(right, p.RightConditions...)
 	p.OtherConditions = append(other, p.OtherConditions...)
 }
 
-func (p *LogicalJoin) extractCorrelatedCols() []*expression.CorrelatedColumn {
-	corCols := p.baseLogicalPlan.extractCorrelatedCols()
+// ExtractCorrelatedCols implements LogicalPlan interface.
+func (p *LogicalJoin) ExtractCorrelatedCols() []*expression.CorrelatedColumn {
+	corCols := make([]*expression.CorrelatedColumn, 0)
```

Comment:
this could introduce a heavy `runtime.growslice` overhead

Reply:
fixed